contracts/DXswapFeeReceiver.sol

```diff
@@ -6,122 +6,226 @@ import './interfaces/IWETH.sol';
+            uint256 feeToAvatarDAO = amount.sub(feeToExternalReceiver);
+            if (token == WETH) {
+                IWETH(WETH).withdraw(amount);
+                TransferHelper.safeTransferETH(_externalFeeReceiver, feeToExternalReceiver);
+                TransferHelper.safeTransferETH(ethReceiver, feeToAvatarDAO);
+            } else {
+                TransferHelper.safeTransfer(token, _externalFeeReceiver, feeToExternalReceiver);
+                TransferHelper.safeTransfer(token, fallbackReceiver, feeToAvatarDAO);
+            }
+        } else {
+            if (token == WETH) {
+                IWETH(WETH).withdraw(amount);
+                TransferHelper.safeTransferETH(ethReceiver, amount);
+            } else {
+                TransferHelper.safeTransfer(token, fallbackReceiver, amount);
+            }
+        }
+    }
+
+    // Convert tokens into ETH if possible, if not just transfer the token
+    function _takeTokenOrETH(
+        address pair,
+        address token,
+        uint256 amount
+    ) internal {
+        if (token != WETH && _isSwapPossible(token, WETH, amount)) {
+            // If it is not WETH and there is a direct path to WETH, swap tokens
+            uint256 amountOut = _swapTokensForETH(amount, token);
+            _splitAndTransferFee(pair, WETH, amountOut);
+        } else {
+            // If it is WETH or there is not a direct path from token to WETH, transfer tokens
+            _splitAndTransferFee(pair, token, amount);
+        }
     }
-    
+
     // Take what was charged as protocol fee from the DXswap pair liquidity
     function takeProtocolFee(IDXswapPair[] calldata pairs) external {
-        for (uint i = 0; i < pairs.length; i++) {
+        for (uint256 i = 0; i < pairs.length; i++) {
             address token0 = pairs[i].token0();
             address token1 = pairs[i].token1();
             pairs[i].transfer(address(pairs[i]), pairs[i].balanceOf(address(this)));
-            (uint amount0, uint amount1) = pairs[i].burn(address(this));
-            if (amount0 > 0)
-                _takeETHorToken(token0, amount0);
-            if (amount1 > 0)
-                _takeETHorToken(token1, amount1);
+            (uint256 amount0, uint256 amount1) = pairs[i].burn(address(this));
+            if (amount0 > 0) _takeTokenOrETH(address(pairs[i]), token0, amount0);
+            if (amount1 > 0) _takeTokenOrETH(address(pairs[i]), token1, amount1);
         }
+        emit TakeProtocolFee(msg.sender, ethReceiver, pairs.length);
+    }
+
+    // called by the owner to set maximum swap price impact allowed for single token-weth swap (within 0-100% range)
+    function setMaxSwapPriceImpact(uint32 _maxSwapPriceImpact) external {
+        require(msg.sender == owner, 'DXswapFeeReceiver: CALLER_NOT_OWNER');
+        require(_maxSwapPriceImpact > 0 && _maxSwapPriceImpact < 10000, 'DXswapFeeReceiver: FORBIDDEN_PRICE_IMPACT');
```

Comment:
The first check `_maxSwapPriceImpact > 0` is not needed since we're dealing with unsigned integers.

Reply:
Indeed it's not needed for unsigned integers, but this check is done to not to set `priceImpact` to zero.